#!/usr/bin/env bash
set -eo pipefail

# Authors:
#   - Arun Manoharan (arun@primediscoveries.com)
# Last Updated February 2019 by Arun Manoharan

# This script generates picrust2 metagenome function prediction tables from an
# ASV feature table and representative sequences. These metagenome prediction
# tables can be used in downstream scripts to produce report visualizations.
#
# The input to this script should be a directory generated by merge_samples.py.
# This script is not intended to be run on single samples at a time prior to
# merging; see the following link for details:
#
# https://forum.qiime2.org/t/merging-picrust2-feature-tables/8064
#
# The script uses the q2-fragment-insertion and q2-picrust2 plugins to insert
# ASV representative sequences into the picrust2 default reference tree and
# predict metagenome functions. See the q2-picrust2 tutorial for details:
#
# https://github.com/picrust/picrust2/wiki/q2-picrust2-Tutorial
#
# The script produces the following tables:
#
# - KO metagenome predictions
# - EC metagenome predictions
# - MetaCyc pathway abundance predictions
#
# Note: the following picrust2 default reference alignment and tree files
# *must* be used with this script (REF_ALIGNMENT and REF_TREE options,
# respectively):
#
# http://kronos.pharmacology.dal.ca/public_files/tutorial_datasets/picrust2_tutorial_files/reference.fna.qza
# http://kronos.pharmacology.dal.ca/public_files/tutorial_datasets/picrust2_tutorial_files/reference.tre.qza

# TODO
#
# - Support other reference databases in addition to the picrust2 default
#   reference files (which are 16S only).

# Requirements:
#
# - MERGED_DIR must contain the output generated by merge_samples.py.
#
# - MERGED_DIR must contain ASVs generated from 16S data because picrust2
#   default reference files are specific to 16S. This requirement may no longer
#   apply if custom picrust2 reference files are supported in the future.

# Program requirements:
#     qiime2-2019.4 conda environment (use install/create_conda_envs.sh to create one)

# Script options:
#     --merged_dir MERGED_DIR: path to directory generated by merge_samples.py [REQUIRED]
#     --ref_alignment REF_ALIGNMENT: path to picrust2 default reference alignment in .qza format [REQUIRED]
#     --ref_tree REF_TREE: path to picrust2 default reference tree in .qza format [REQUIRED]
#     --threads THREADS: number of threads to use for parallel processes [default: 8]
#     --qiime2_env QIIME2_ENV: qiime2 conda environment name [default: 'qiime2-2019.4']

# Outputs:
#     ${MERGED_DIR}/picrust_predictions/
#         picrust_predictions.log: log of commands started and completed during running of this script
#         1_sepp_placement: directory containing ASV representative sequences placed into picrust2 reference tree using SEPP
#         2_metagenome_prediction: directory containing predicted metagenome feature tables

# Example usage:
#     bash picrust_predictions.sh
#         --merged_dir merged_samples
#         --ref_alignment reference.fna.qza
#         --ref_tree reference.tre.qza
#         --threads 8

MERGED_DIR=""
REF_ALIGNMENT=""
REF_TREE=""
THREADS="8"
QIIME2_ENV="qiime2-2019.4"

# https://stackoverflow.com/a/14203146/3776794
while [[ $# -gt 0 ]]; do
    key="${1}"

    case "${key}" in
        --merged_dir)
            MERGED_DIR="${2}"
            shift 2
            ;;
        --ref_alignment)
            REF_ALIGNMENT="${2}"
            shift 2
            ;;
        --ref_tree)
            REF_TREE="${2}"
            shift 2
            ;;
        --threads)
            THREADS="${2}"
            shift 2
            ;;
        --qiime2_env)
            QIIME2_ENV="${2}"
            shift 2
            ;;
        *)
            echo "Error: Unknown option '${key}'. Please see script header for usage instructions."
            exit 1
            ;;
    esac
done

if [ -z "${MERGED_DIR}" ]; then
    echo "Error: --merged_dir is a required option."
    exit 1
fi

if [ -z "${REF_ALIGNMENT}" ]; then
    echo "Error: --ref_alignment is a required option."
    exit 1
fi

if [ -z "${REF_TREE}" ]; then
    echo "Error: --ref_tree is a required option."
    exit 1
fi

OUTPUT_DIR="${MERGED_DIR}/picrust_predictions"

if [ -d "${OUTPUT_DIR}" ]; then
  echo "Error: ${OUTPUT_DIR} directory already exists. Please either remove the directory and rerun this script or specify a different --merged_dir."
  exit 1
fi

mkdir -p ${OUTPUT_DIR}

# Writes message with timestamp to log file and also displays on stdout. Will
# create log file if it doesn't already exist.
function log_message {
    set -eo pipefail

    MSG="${1}"
    SCRIPT_LOG="${OUTPUT_DIR}/picrust_predictions.log"

    if [ ! -f "${SCRIPT_LOG}" ]; then
        echo -e "picrust_predictions.sh log file\n" > "${SCRIPT_LOG}"
    fi

    echo "$(date +%Y-%m-%d:%H:%M:%S) ${MSG}" | tee -a "${SCRIPT_LOG}"
}

# Save pipeline start timestamp for writing to pipeline_completed.log file at
# end of script.
PIPELINE_STARTED="$(date +%Y-%m-%d:%H:%M:%S)"
log_message "Pipeline execution started."

# Create directories for output
log_message "Creating output directories started."

mkdir -p ${OUTPUT_DIR}/1_sepp_placement
mkdir -p ${OUTPUT_DIR}/2_metagenome_prediction

log_message "Creating output directories completed."

# export a copy of input parameters
log_message "Writing file of input parameters started."
PARAMS_LOG="${OUTPUT_DIR}/picrust_predictions_params.log"

echo "MERGED_DIR: $MERGED_DIR" > "${PARAMS_LOG}"
echo "REF_ALIGNMENT: $REF_ALIGNMENT" >> "${PARAMS_LOG}"
echo "REF_TREE: $REF_TREE" >> "${PARAMS_LOG}"
echo "THREADS: $THREADS" >> "${PARAMS_LOG}"
echo "QIIME2_ENV: $QIIME2_ENV" >> "${PARAMS_LOG}"

log_message "Writing file of input parameters completed."

source activate ${QIIME2_ENV}

log_message "Placing ASV representative sequences into picrust2 default reference tree with SEPP started."

REP_SEQS_PATH="${MERGED_DIR}/asvs/rep_seqs.qza"

# The picrust2 default reference alignment and tree *must* be used here; the
# default SEPP reference files won't work.
qiime fragment-insertion sepp \
    --i-representative-sequences ${REP_SEQS_PATH} \
    --i-reference-alignment ${REF_ALIGNMENT} \
    --i-reference-phylogeny ${REF_TREE} \
    --p-threads ${THREADS} \
    --o-tree ${OUTPUT_DIR}/1_sepp_placement/insertion_tree.qza \
    --o-placements ${OUTPUT_DIR}/1_sepp_placement/insertion_placements.qza \
    --verbose \
    2>&1 | tee -a ${OUTPUT_DIR}/q2_fragment_insertion_sepp.log

log_message "Placing ASV representative sequences into picrust2 default reference tree with SEPP completed."

log_message "Predicting metagenomes with picrust2 started."

FEATURE_TABLE_PATH="${MERGED_DIR}/asvs/feature_table.qza"

# hsp-method and max-nsti parameters chosen based on picrust2 tutorial recommendations:
# https://github.com/picrust/picrust2/wiki/q2-picrust2-Tutorial
#
# Note that max-nsti=2 works well for human data sets, may need to change for environmental data sets.
qiime picrust2 custom-tree-pipeline \
    --i-table ${FEATURE_TABLE_PATH} \
    --i-tree ${OUTPUT_DIR}/1_sepp_placement/insertion_tree.qza \
    --p-threads ${THREADS} \
    --p-hsp-method mp \
    --p-max-nsti 2 \
    --o-ko-metagenome ${OUTPUT_DIR}/2_metagenome_prediction/ko_metagenome.qza \
    --o-ec-metagenome ${OUTPUT_DIR}/2_metagenome_prediction/ec_metagenome.qza \
    --o-pathway-abundance ${OUTPUT_DIR}/2_metagenome_prediction/pathway_abundance.qza \
    --verbose \
    2>&1 | tee -a ${OUTPUT_DIR}/q2_picrust2_custom_tree_pipeline.log

log_message "Predicting metagenomes with picrust2 completed."

# Deactivate qiime2 conda env.
conda deactivate

PIPELINE_COMPLETED="$(date +%Y-%m-%d:%H:%M:%S)"

echo "${PIPELINE_STARTED} Pipeline execution started." > "${OUTPUT_DIR}/pipeline_completed.log"
echo "${PIPELINE_COMPLETED} Pipeline execution completed." >> "${OUTPUT_DIR}/pipeline_completed.log"
log_message "Pipeline execution completed."
